src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler/Deprecation.php

```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    private function lineShouldBeSkipped(array $line)
+    {
+        if (!isset($line['class'])) {
+            return true;
+        }
+        $class = $line['class'];
+
+        return 'ReflectionMethod' === $class || 0 === strpos($class, 'PHPUnit_') || 0 === strpos($class, 'PHPUnit\\');
+    }
+
+    /**
+     * @return bool
+     */
+    public function originatesFromAnObject()
+    {
+        return isset($this->originClass);
+    }
+
+    /**
+     * @return bool
+     */
+    public function isSelf()
+    {
+        return $this->self;
+    }
+
+    /**
+     * @return string
+     */
+    public function originatingClass()
+    {
+        if (null === $this->originClass) {
+            throw new \LogicException('Check with originatesFromAnObject() before calling this method');
+        }
+
+        return $this->originClass;
+    }
+
+    /**
+     * @return string
+     */
+    public function originatingMethod()
+    {
+        if (null === $this->originMethod) {
+            throw new \LogicException('Check with originatesFromAnObject() before calling this method');
+        }
+
+        return $this->originMethod;
+    }
+
+    /**
+     * @param string $utilPrefix
+     *
+     * @return bool
+     */
+    public function isLegacy($utilPrefix)
+    {
+        $test = $utilPrefix.'Test';
+        $class = $this->originatingClass();
+        $method = $this->originatingMethod();
+
+        return 0 === strpos($method, 'testLegacy')
+            || 0 === strpos($method, 'provideLegacy')
+            || 0 === strpos($method, 'getLegacy')
+            || strpos($class, '\Legacy')
+            || \in_array('legacy', $test::getGroups($class, $method), true);
+    }
+
+    /**
+     * Tells whether both the calling package and the called package are vendor
+     * packages.
+     *
+     * @return bool
+     */
+    public function isIndirect()
+    {
+        $erroringFile = $erroringPackage = null;
+        foreach ($this->trace as $line) {
+            if (\in_array($line['function'], ['require', 'require_once', 'include', 'include_once'], true)) {
+                continue;
+            }
+            if (!isset($line['file'])) {
+                continue;
+            }
+            $file = $line['file'];
+            if ('-' === $file || 'Standard input code' === $file || !realpath($file)) {
+                continue;
+            }
+            if (!$this->pathOriginatesFromVendor($file)) {
+                return false;
+            }
+            if (null !== $erroringFile && null !== $erroringPackage) {
+                $package = $this->getPackage($file);
+                if ('composer' !== $package && $package !== $erroringPackage) {
+                    return true;
+                }
+                continue;
+            }
+            $erroringFile = $file;
+            $erroringPackage = $this->getPackage($file);
+        }
+
+        return false;
+    }
```

Comment:
@greg0ire what is the goal of this method please ? Is it only used for the color of the deprecation text or I missed another usage ? I saw some minors problems (see #31478 (WIP)). When using `runInSeparateProcess`, it's not the correct stack trace that is used to determine `isIndirect` result. I'm not sure how fix it. In some cases, I will not be able to serialize the deprecation original stack trace (when there is a closure). What would be (from your point of view) the consequences if isIndirect can't be determined ?

Reply:
The goal is to help making a difference between direct and indirect. If the result is inaccurate, a test suite might fail when it should not and vice versa. I'm on hols, and answering from my phone while walking in the street, and GH doesn't show the name of the method so my answer might be inaccurate too :P

Reply:
IIRC there is a group of "others" deprecations, maybe we should throw an exception when unable to answer, and catch it and then put the deprecation in that group

Reply:
I thought too about this kind of solution. To accept to not be able to determine direct/indirect in some edge cases (runInSeparateProcess and unserializable trace in this case). Thank you for your help 👍

Reply:
See  implementation of this kind of solution here: https://github.com/symfony/symfony/pull/31730